File: languages/php/Bloock/Client/IdentityClient.php
<?php

namespace Bloock\Client;

use Bloock\Bridge\Bridge;
use Bloock\Config\Config;
use Bloock\ConfigData;
use Bloock\CreateHolderRequest;
use Bloock\CreateIssuerRequest;
use Bloock\CreateVerificationRequest;
use Bloock\Entity\Identity\DidType;
use Bloock\Entity\Identity\PublishIntervalParams;
use Bloock\Entity\Identity\Credential;
use Bloock\Entity\Identity\Schema;
use Bloock\Entity\Identity\CredentialBuilder;
use Bloock\Entity\Identity\CredentialProof;
use Bloock\Entity\Identity\DidMethod;
use Bloock\Entity\Identity\Issuer;
use Bloock\Entity\Identity\Holder;
use Bloock\Entity\Identity\IssuerStateReceipt;
use Bloock\Entity\Identity\SchemaBuilder;
use Bloock\Entity\Identity\VerificationReceipt;
use Bloock\Entity\Key\Key;
use Bloock\GetCredentialProofRequest;
use Bloock\GetSchemaRequest;
use Bloock\GetVerificationStatusRequest;
use Bloock\ImportIssuerRequest;
use Bloock\ForcePublishIssuerStateRequest;
use Bloock\GetCredentialOfferRequest;
use Bloock\GetCredentialRequest;
use Bloock\RevokeCredentialRequest;
use Bloock\WaitVerificationRequest;
use Exception;

/**
 * Represents a client for interacting with the [Bloock Identity service](https://dashboard.bloock.com/login).
 */
class IdentityClient
{
    private $bridge;
    private $config;

    /**
     * Creates a new instance of the IdentityClient with the provided configuration.
     * @param ConfigData|null $config
     */
    public function __construct(ConfigData $config = null)
    {
        $this->bridge = new Bridge();
        if ($config != null) {
            $this->config = Config::newConfigData($config);
        } else {
            $this->config = Config::newConfigDataDefault();
        }
    }

    /**
     * Creates a new holder identity.
     * @param Key $holderKey
     * @param string $didMethod
     * @return Holder
     * @throws Exception
     */
    public function createHolder(Key $holderKey, string $didMethod): Holder
    {
        $req = new CreateHolderRequest();
        $req->setKey($holderKey->toProto());
        $req->setConfigData($this->config);
        $req->setDidMethod(DidMethod::toProto($didMethod));

        $res = $this->bridge->identity->CreateHolder($req);

        if ($res->getError() != null) {
            throw new Exception($res->getError()->getMessage());
        }

        return new Holder($res->getDid(), new DidMethod($didMethod), $holderKey);
    }

    /**
     * Creates a new issuer on the Bloock Identity service.
     * @param Key $issuerKey
     * @param int $publishInterval
     * @param string $didMethod
     * @param string|null $name
     * @param string|null $description
     * @param string|null $image
     * @return Issuer
     * @throws Exception
     */
    public function createIssuer(Key $issuerKey, int $publishInterval, string $didMethod, string $name = null, string $description = null, string $image = null): Issuer
    {
        $req = new CreateIssuerRequest();
        $req->setKey($issuerKey->toProto());
        $req->setPublishInterval(PublishIntervalParams::toProto($publishInterval));
        $req->setConfigData($this->config);
        $req->setDidMethod(DidMethod::toProto($didMethod));

        if ($name != null) {
            $req->setName($name);
        }

        if ($description != null) {
            $req->setDescription($description);
        }

        if ($image != null) {
            $req->setImage($image);
        }

        $res = $this->bridge->identity->CreateIssuer($req);

        if ($res->getError() != null) {
            throw new Exception($res->getError()->getMessage());
        }

        return new Issuer($res->getDid(), new DidMethod($didMethod), $issuerKey);
    }

    /**
     * Gets the issuer based on the issuer key and DID method.
     * @param Key $issuerKey
     * @param string $didMethod
     * @return Issuer
     * @throws Exception
     */
    public function importIssuer(Key $issuerKey, string $didMethod): Issuer
    {
        $req = new ImportIssuerRequest();
        $req->setKey($issuerKey->toProto());
        $req->setConfigData($this->config);
        $req->setDidMethod(DidMethod::toProto($didMethod));

        $res = $this->bridge->identity->ImportIssuer($req);

        if ($res->getError() != null) {
            throw new Exception($res->getError()->getMessage());
        }

        return new Issuer($res->getDid(), new DidMethod($didMethod), $issuerKey);
    }

    /**
     * Creates a new schema builder for defining a schema on the Bloock Identity service.
     * @param string $displayName
     * @param string $schemaType
     * @param string $version
     * @param string $description
     * @return SchemaBuilder
     */
    public function buildSchema(string $displayName, string $schemaType, string $version, string $description): SchemaBuilder
    {
        return new SchemaBuilder($displayName, $schemaType, $version, $description, $this->config);
    }

    /**
     * Gets a schema from the Bloock Identity service based on the schema ID (ex: Qma1t4uzbnB93E4rasNdu5UWMDh5qg3wMkPm68cnEyfnoM).
     * @param string $id
     * @return Schema
     * @throws Exception
     */
    public function getSchema(string $id): Schema
    {
        $req = new GetSchemaRequest();
        $req->setConfigData($this->config)->setId($id);

        $res = $this->bridge->identity->GetSchema($req);

        if ($res->getError() != null) {
            throw new Exception($res->getError()->getMessage());
        }

        return Schema::fromProto($res->getSchema());
    }

    /**
     * Creates a new credential builder for defining a credential on the Bloock Identity service.
     * @param Issuer $issuer
     * @param string $schemaId
     * @param string $holderDid
     * @param int $expiration
     * @param int $version
     * @return CredentialBuilder
     */
    public function buildCredential(Issuer $issuer, string $schemaId, string $holderDid, int $expiration, int $version): CredentialBuilder
    {
        return new CredentialBuilder($issuer, $schemaId, $holderDid, $expiration, $version, $this->config);
    }

    /**
     * Retrieves the Verifiable Credential entity based on the credential ID (UUID). (ex: 1bf0c79e-55e6-4f14-aa9d-fb55619ba0cf)
     * @param string $credentialId
     * @return Credential
     */
    public function getCredential(string $credentialId): Credential
    {
        $req = new GetCredentialRequest();
        $req->setConfigData($this->config)->setCredentialId($credentialId);

        $res = $this->bridge->identity->GetCredential($req);

        if ($res->getError() != null) {
            throw new Exception($res->getError()->getMessage());
        }

        return Credential::fromProto($res->getCredential());
    }

    /**
     * Retrieves the json raw offer based on the credential ID (UUID). (ex: 1bf0c79e-55e6-4f14-aa9d-fb55619ba0cf)
     * @param Issuer $issuer
     * @param string $credentialId
     */
    public function getCredentialOffer(Issuer $issuer, string $credentialId): string
    {
        $req = new GetCredentialOfferRequest();
        $req->setConfigData($this->config)->setCredentialId($credentialId)->setKey($issuer->getKey()->toProto());

        $res = $this->bridge->identity->GetCredentialOffer($req);

        if ($res->getError() != null) {
            throw new Exception($res->getError()->getMessage());
        }

        return $res->getCredentialOffer();
    }

    /**
     * Publishes the state of an issuer on the Bloock Identity service.
     * @param Issuer $issuer
     * @return IssuerStateReceipt
     * @throws Exception
     */
    public function forcePublishIssuerState(Issuer $issuer): IssuerStateReceipt
    {
        $req = new ForcePublishIssuerStateRequest();
        $req->setConfigData($this->config);
        $req->setIssuerDid($issuer->getDid()->getDid());
        $req->setKey($issuer->getKey()->toProto());

        $res = $this->bridge->identity->ForcePublishIssuerState($req);

        if ($res->getError() != null) {
            throw new Exception($res->getError()->getMessage());
        }

        return IssuerStateReceipt::fromProto($res->getStateReceipt());
    }

    /**
     * Gets the proof of a credential on the Bloock Identity service.
     * @param string $issuerDid
     * @param string $credentialId
     * @return CredentialProof
     * @throws Exception
     */
    public function getCredentialProof(string $issuerDid, string $credentialId): CredentialProof
    {
        $req = new GetCredentialProofRequest();
        $req->setIssuerDid($issuerDid);
        $req->setCredentialId($credentialId);
        $req->setConfigData($this->config);

        $res = $this->bridge->identity->getCredentialProof($req);

        if ($res->getError() != null) {
            throw new Exception($res->getError()->getMessage());
        }

        return CredentialProof::fromProto($res->getProof());
    }

    /**
     * Revokes a credential on the Bloock Identity service.
     * @param Credential $credential
     * @param Issuer $issuer
     * @return bool
     * @throws Exception
     */
    public function revokeCredential(Credential $credential, Issuer $issuer): bool
    {
        $req = new RevokeCredentialRequest();
        $req->setConfigData($this->config);
        $req->setCredential($credential->toProto());
        $req->setKey($issuer->getKey()->toProto());

        $res = $this->bridge->identity->RevokeCredential($req);

        if ($res->getError() != null) {
            throw new Exception($res->getError()->getMessage());
        }

        return $res->getResult()->getSuccess();
    }

    /**
     * Creates a new verification session on the identity managed API provided.
     * @param String $proofRequest
     * @return VerificationReceipt
     * @throws Exception
     */
    public function createVerification(String $proofRequest): VerificationReceipt
    {
        $req = new CreateVerificationRequest();
        $req->setConfigData($this->config);
        $req->setProofRequest($proofRequest);

        $res = $this->bridge->identity->CreateVerification($req);

        if ($res->getError() != null) {
            throw new Exception($res->getError()->getMessage());
        }

        return VerificationReceipt::fromProto($res->getResult());
    }

    /**
     * Waits for the completion of a verification session on the identity managed API provided.
     * @param int $sessionID
     * @param int $timeout
     * @return bool
     * @throws Exception
     */
    public function waitVerification(int $sessionID, int $timeout = 120000): bool
    {
        $req = new WaitVerificationRequest();
        $req->setConfigData($this->config);
        $req->setSessionId($sessionID);
        $req->setTimeout($timeout);

        $res = $this->bridge->identity->WaitVerification($req);

        if ($res->getError() != null) {
            throw new Exception($res->getError()->getMessage());
        }

        return $res->getStatus();
    }

    /**
     * Gets the status of a verification session on the identity managed API provided.
     * @param int $sessionID
     * @return bool
     * @throws Exception
     */
    public function getVerificationStatus(int $sessionID): bool
    {
        $req = new GetVerificationStatusRequest();
        $req->setConfigData($this->config);
        $req->setSessionId($sessionID);

        $res = $this->bridge->identity->GetVerificationStatus($req);

        if ($res->getError() != null) {
            throw new Exception($res->getError()->getMessage());
        }

        return $res->getStatus();
    }
}

File: languages/java/src/main/java/com/bloock/sdk/client/IdentityClient.java
package com.bloock.sdk.client;

import com.bloock.sdk.bridge.Bridge;
import com.bloock.sdk.bridge.proto.BloockConfig.ConfigData;
import com.bloock.sdk.bridge.proto.BloockIdentity;
import com.bloock.sdk.bridge.proto.BloockShared;
import com.bloock.sdk.bridge.proto.BloockShared.Error;
import com.bloock.sdk.config.Config;
import com.bloock.sdk.entity.identity.*;
import com.bloock.sdk.entity.key.Key;

/**
 * Represents a client for interacting with the
 * <a href="https://dashboard.bloock.com/login">Bloock Identity service</a>.
 */
public class IdentityClient {
  private final Bridge bridge;
  private final ConfigData configData;

  /**
   * Creates a new instance of the IdentityClient with default configuration.
   */
  public IdentityClient() {
    this.bridge = new Bridge();
    this.configData = Config.newConfigDataDefault();
  }

  /**
   * Creates a new instance of the IdentityClient with the provided configuration.
   * 
   * @param configData
   */
  public IdentityClient(ConfigData configData) {
    this.bridge = new Bridge();
    this.configData = Config.newConfigData(configData);
  }

  /**
   * Creates a new holder identity.
   * 
   * @param holderKey
   * @param didMethod
   * @return
   * @throws Exception
   */
  public Holder createHolder(Key holderKey, DidMethod didMethod) throws Exception {
    BloockIdentity.CreateHolderRequest.Builder builder = BloockIdentity.CreateHolderRequest.newBuilder()
        .setKey(holderKey.toProto())
        .setDidMethod(didMethod.toProto())
        .setConfigData(this.configData);

    BloockIdentity.CreateHolderRequest request = builder.build();

    BloockIdentity.CreateHolderResponse response = bridge.getIdentity().createHolder(request);

    if (response.getError() != Error.getDefaultInstance()) {
      throw new Exception(response.getError().getMessage());
    }

    return new Holder(response.getDid(), didMethod, holderKey);
  }

  /**
   * Creates a new issuer on the Bloock Identity service.
   * 
   * @param issuerKey
   * @param publishInterval
   * @param didMethod
   * @param name
   * @param description
   * @param image
   * @return
   * @throws Exception
   */
  public Issuer createIssuer(
      Key issuerKey,
      PublishIntervalParams publishInterval,
      DidMethod didMethod,
      String name,
      String description,
      String image)
      throws Exception {
    BloockIdentity.CreateIssuerRequest.Builder builder = BloockIdentity.CreateIssuerRequest.newBuilder()
        .setKey(issuerKey.toProto())
        .setDidMethod(didMethod.toProto())
        .setPublishInterval(publishInterval.toProto())
        .setConfigData(this.configData);

    if (name != null) {
      builder.setName(name);
    }

    if (description != null) {
      builder.setDescription(description);
    }

    if (image != null) {
      builder.setImage(image);
    }

    BloockIdentity.CreateIssuerRequest request = builder.build();

    BloockIdentity.CreateIssuerResponse response = bridge.getIdentity().createIssuer(request);

    if (response.getError() != Error.getDefaultInstance()) {
      throw new Exception(response.getError().getMessage());
    }

    return new Issuer(response.getDid(), didMethod, issuerKey);
  }

  /**
   * Gets the issuer based on the issuer key and DID method.
   * 
   * @param issuerKey
   * @param didMethod
   * @return
   * @throws Exception
   */
  public Issuer importIssuer(Key issuerKey, DidMethod didMethod) throws Exception {
    BloockIdentity.ImportIssuerRequest request = BloockIdentity.ImportIssuerRequest.newBuilder()
        .setKey(issuerKey.toProto())
        .setDidMethod(didMethod.toProto())
        .setConfigData(this.configData)
        .build();

    BloockIdentity.ImportIssuerResponse response = bridge.getIdentity().importIssuer(request);

    if (response.getError() != Error.getDefaultInstance()) {
      throw new Exception(response.getError().getMessage());
    }

    return new Issuer(response.getDid(), didMethod, issuerKey);
  }

  /**
   * Creates a new schema builder for defining a schema on the Bloock Identity
   * service.
   * 
   * @param displayName
   * @param schemaType
   * @param version
   * @param description
   * @return
   * @throws Exception
   */
  public SchemaBuilder buildSchema(
      String displayName, String schemaType, String version, String description) throws Exception {
    return new SchemaBuilder(displayName, schemaType, version, description, this.configData);
  }

  /**
   * Gets a schema from the Bloock Identity service based on the schema ID (ex:
   * Qma1t4uzbnB93E4rasNdu5UWMDh5qg3wMkPm68cnEyfnoM).
   * 
   * @param id
   * @return
   * @throws Exception
   */
  public Schema getSchema(String id) throws Exception {
    BloockIdentity.GetSchemaRequest request = BloockIdentity.GetSchemaRequest.newBuilder()
        .setConfigData(this.configData).setId(id)
        .build();

    BloockIdentity.GetSchemaResponse response = bridge.getIdentity().getSchema(request);

    if (response.getError() != Error.getDefaultInstance()) {
      throw new Exception(response.getError().getMessage());
    }

    return Schema.fromProto(response.getSchema());
  }

  /**
   * Creates a new credential builder for defining a credential on the Bloock
   * Identity service.
   * 
   * @param issuer
   * @param schemaId
   * @param holderDid
   * @param expiration
   * @param version
   * @return
   * @throws Exception
   */
  public CredentialBuilder buildCredential(
      Issuer issuer, String schemaId, String holderDid, Long expiration, int version)
      throws Exception {
    return new CredentialBuilder(
        issuer, schemaId, holderDid, expiration, version, this.configData);
  }

  /**
   * Retrieves the Verifiable Credential entity based on the credential ID (UUID).
   * (ex: 1bf0c79e-55e6-4f14-aa9d-fb55619ba0cf)
   * 
   * @param credentialId
   * @return
   * @throws Exception
   */
  public Credential getCredential(String credentialId) throws Exception {
    BloockIdentity.GetCredentialRequest request = BloockIdentity.GetCredentialRequest.newBuilder()
        .setConfigData(this.configData)
        .setCredentialId(credentialId).build();

    BloockIdentity.GetCredentialResponse response = bridge.getIdentity().getCredential(request);

    if (response.getError() != Error.getDefaultInstance()) {
      throw new Exception(response.getError().getMessage());
    }

    return Credential.fromProto(response.getCredential());
  }

  /**
   * Retrieves the json raw offer based on the credential ID (UUID). (ex:
   * 1bf0c79e-55e6-4f14-aa9d-fb55619ba0cf)
   * 
   * @param issuer
   * @param credentialId
   * @return
   * @throws Exception
   */
  public String getCredentialOffer(Issuer issuer, String credentialId) throws Exception {
    BloockIdentity.GetCredentialOfferRequest request = BloockIdentity.GetCredentialOfferRequest.newBuilder()
        .setConfigData(this.configData).setCredentialId(credentialId).setKey(issuer.getKey().toProto()).build();

    BloockIdentity.GetCredentialOfferResponse response = bridge.getIdentity().getCredentialOffer(request);

    if (response.getError() != Error.getDefaultInstance()) {
      throw new Exception(response.getError().getMessage());
    }

    return response.getCredentialOffer();
  }

  /**
   * Publishes the state of an issuer on the Bloock Identity service.
   * 
   * @param issuer
   * @return
   * @throws Exception
   */
  public IssuerStateReceipt forcePublishIssuerState(Issuer issuer) throws Exception {
    BloockIdentity.ForcePublishIssuerStateRequest req = BloockIdentity.ForcePublishIssuerStateRequest.newBuilder()
        .setConfigData(this.configData)
        .setIssuerDid(issuer.getDid().getDid())
        .setKey(issuer.getKey().toProto())
        .build();

    BloockIdentity.ForcePublishIssuerStateResponse response = bridge.getIdentity().forcePublishIssuerState(req);

    if (response.getError() != BloockShared.Error.getDefaultInstance()) {
      throw new Exception(response.getError().getMessage());
    }

    return IssuerStateReceipt.fromProto(response.getStateReceipt());
  }

  /**
   * Gets the proof of a credential on the Bloock Identity service.
   * 
   * @param issuerDid
   * @param credentialId
   * @return
   * @throws Exception
   */
  public CredentialProof getCredentialProof(String issuerDid, String credentialId)
      throws Exception {
    BloockIdentity.GetCredentialProofRequest request = BloockIdentity.GetCredentialProofRequest.newBuilder()
        .setCredentialId(credentialId)
        .setIssuerDid(issuerDid)
        .setConfigData(this.configData)
        .build();

    BloockIdentity.GetCredentialProofResponse response = bridge.getIdentity().getCredentialProof(request);

    if (response.getError() != Error.getDefaultInstance()) {
      throw new Exception(response.getError().getMessage());
    }

    return CredentialProof.fromProto(response.getProof());
  }

  /**
   * Revokes a credential on the Bloock Identity service.
   * 
   * @param credential
   * @param issuer
   * @return
   * @throws Exception
   */
  public boolean revokeCredential(Credential credential, Issuer issuer) throws Exception {
    BloockIdentity.RevokeCredentialRequest request = BloockIdentity.RevokeCredentialRequest.newBuilder()
        .setConfigData(this.configData)
        .setCredential(credential.toProto())
        .setKey(issuer.getKey().toProto())
        .build();

    BloockIdentity.RevokeCredentialResponse response = bridge.getIdentity().revokeCredential(request);

    if (response.getError() != Error.getDefaultInstance()) {
      throw new Exception(response.getError().getMessage());
    }

    return response.getResult().getSuccess();
  }

  /**
   * Creates a new verification session on the identity managed API provided.
   * 
   * @param proofRequest
   * @return
   * @throws Exception
   */
  public VerificationReceipt createVerification(String proofRequest) throws Exception {
    BloockIdentity.CreateVerificationRequest request = BloockIdentity.CreateVerificationRequest.newBuilder()
        .setConfigData(this.configData)
        .setProofRequest(proofRequest)
        .build();

    BloockIdentity.CreateVerificationResponse response = bridge.getIdentity().createVerification(request);

    if (response.getError() != Error.getDefaultInstance()) {
      throw new Exception(response.getError().getMessage());
    }

    return VerificationReceipt.fromProto(response.getResult());
  }

  /**
   * Waits for the completion of a verification session on the identity managed
   * API provided.
   * 
   * @param sessionID
   * @param timeout
   * @return
   * @throws Exception
   */
  public boolean waitVerification(long sessionID, long timeout) throws Exception {
    BloockIdentity.WaitVerificationRequest request = BloockIdentity.WaitVerificationRequest.newBuilder()
        .setConfigData(this.configData)
        .setSessionId(sessionID)
        .setTimeout(timeout)
        .build();

    BloockIdentity.WaitVerificationResponse response = bridge.getIdentity().waitVerification(request);

    if (response.getError() != Error.getDefaultInstance()) {
      throw new Exception(response.getError().getMessage());
    }

    return response.getStatus();
  }

  /**
   * Gets the status of a verification session on the identity managed API
   * provided.
   * 
   * @param sessionID
   * @return
   * @throws Exception
   */
  public boolean getVerificationStatus(long sessionID) throws Exception {
    BloockIdentity.GetVerificationStatusRequest request = BloockIdentity.GetVerificationStatusRequest.newBuilder()
        .setConfigData(this.configData)
        .setSessionId(sessionID)
        .build();

    BloockIdentity.GetVerificationStatusResponse response = bridge.getIdentity().getVerificationStatus(request);

    if (response.getError() != Error.getDefaultInstance()) {
      throw new Exception(response.getError().getMessage());
    }

    return response.getStatus();
  }
}
